dockerfile-image-update/src/test/java/com/salesforce/dockerfileimageupdate/utils/RateLimiterTest.java

```diff
@@ -0,0 +1,139 @@
+    }
+
+    /**
+     * Using Custom Clock to increment clock counter
+     * Instead of 60 seconds this should take < second to get completed
+     *
+     * @throws Exception
+     */
+    @Test(dataProvider = "rateLimitingData")
+    public void testingAgainstCustomClock(int rateLimit, int durationLimit, int tokenAddingRate) throws Exception {
+        RateLimiterEvent event = new RateLimiterEvent
+                (new CustomTimeMeter(), rateLimit, durationLimit, tokenAddingRate);
+        FutureTask futureTask = new FutureTask(new RateLimiterCallable(event));
+        Thread t1 = new Thread(futureTask);
+        long startTime = System.currentTimeMillis();
+        t1.start();
+        t1.join();
+        long endTime = System.currentTimeMillis();
+        assertTrue((endTime - startTime) < (tokenAddingRate * 1_000L));
+        assertEquals(event.getProcessedEventCount(), rateLimit);
+        assertEquals(t1.getState().toString(), "TERMINATED");
+        assertEquals(futureTask.get(), Optional.ofNullable(null));
+    }
+}
+
+
+class RateLimiterCallable implements Callable {
+    RateLimiterEvent rateLimiterEvent;
+
+    RateLimiterCallable(RateLimiterEvent rateLimiterEvent) {
+        this.rateLimiterEvent = rateLimiterEvent;
+    }
+
+    @Test
+    @Override
+    public Optional<Exception> call() {
+        try {
+            RateLimiter rateLimiter = new RateLimiter(rateLimiterEvent.ratelimit, Duration.ofSeconds(rateLimiterEvent.durationLimit),
+                    Duration.ofSeconds(rateLimiterEvent.tokenAddingRate), RateLimiterEvent.clock);
+            int counter = 0;
+            while (counter++ < rateLimiterEvent.ratelimit) {
+                rateLimiter.consume();
+                rateLimiterEvent.incrementEventProcessed();
+                assertEquals(Thread.currentThread().getState().toString(), "RUNNABLE");
+                if (RateLimiterEvent.clock != null) {
+                    RateLimiterEvent.clock.addSeconds(rateLimiterEvent.tokenAddingRate);
+                }
+            }
+            return Optional.ofNullable(null);
+        } catch (InterruptedException e) {
+            System.out.println("exception thrown");
+            return Optional.ofNullable(new InterruptedException());
+        }
+    }
+}
+
+
+class RateLimiterEvent {
+    static CustomTimeMeter clock;
+    long ratelimit;
+    long durationLimit;
+    long tokenAddingRate;
+    volatile int processedEventCount = 0;
+
+    public RateLimiterEvent(CustomTimeMeter clock, long ratelimit, long durationLimit, long tokenAddingRate) {
+        RateLimiterEvent.clock = clock;
+        this.ratelimit = ratelimit;
+        this.durationLimit = durationLimit;
+        this.tokenAddingRate = tokenAddingRate;
+    }
+
+    public void incrementEventProcessed() {
+        this.processedEventCount++;
+    }
+
+    public int getProcessedEventCount() {
+        return processedEventCount;
+    }
+}
+
+class CustomTimeMeter implements TimeMeter {
```

Comment:
Is there a better way to describe this than `Custom`?

Reply:
Used MockTimeMeter.. does that looks okay?